processing/src/main/java/io/druid/segment/IndexMergerV9.java

```diff
@@ -859,72 +779,48 @@ private void writeDimValueAndSetupDimConversion(
       long dimStartTime = System.currentTimeMillis();
       String dimension = mergedDimensions.get(dimIndex);
 
-      // lookups for all dimension values of this dimension
-      List<Indexed<String>> dimValueLookups = Lists.newArrayListWithCapacity(adapters.size());
-
-      // each converter converts dim values of this dimension to global dictionary
-      DimValueConverter[] converters = new DimValueConverter[adapters.size()];
+      Indexed<String> dimValueLookup = null;
+      Indexed<String>[] dimValueLookups = new Indexed[adapters.size() + 1];
 
+      int numMergeIndex = 0;
       boolean existNullColumn = false;
       for (int i = 0; i < adapters.size(); i++) {
         Indexed<String> dimValues = adapters.get(i).getDimValueLookup(dimension);
         if (!isNullColumn(dimValues)) {
-          dimValueLookups.add(dimValues);
-          converters[i] = new DimValueConverter(dimValues);
+          dimValueLookups[i] = dimValueLookup = dimValues;
+          numMergeIndex++;
         } else {
           existNullColumn = true;
         }
       }
 
-      Iterable<Indexed<String>> bumpedDimValueLookups;
-      if (!dimValueLookups.isEmpty() && existNullColumn) {
+      if (numMergeIndex > 0 && existNullColumn) {
         log.info("dim[%s] are null in some indexes, append null value to dim values", dimension);
-        bumpedDimValueLookups = Iterables.concat(
-            Arrays.asList(new ArrayIndexed<>(new String[]{null}, String.class)),
-            dimValueLookups
-        );
-      } else {
-        bumpedDimValueLookups = dimValueLookups;
+        dimValueLookups[adapters.size()] = new ArrayIndexed<>(new String[]{null}, String.class);
+        dimValueLookup = dimValueLookups[adapters.size()];
+        numMergeIndex++;
       }
 
-      // sort all dimension values and treat all null values as empty strings
-      Iterable<String> dimensionValues = CombiningIterable.createSplatted(
-          Iterables.transform(
-              bumpedDimValueLookups,
-              new Function<Indexed<String>, Iterable<String>>()
-              {
-                @Override
-                public Iterable<String> apply(@Nullable Indexed<String> indexed)
-                {
-                  return Iterables.transform(
-                      indexed,
-                      new Function<String, String>()
-                      {
-                        @Override
-                        public String apply(@Nullable String input)
-                        {
-                          return (input == null) ? "" : input;
```

Comment:
can you double check if you remove this code, how many dictionary entries are created for empty string and null?